azure-iot-device/azure/iot/device/iothub/aio/loop_management.py

```diff
@@ -0,0 +1,72 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+""" This module contains functions of managing event loops for the IoTHub client
+"""
+import asyncio
+import threading
+import logging
+from azure.iot.device.common import asyncio_compat
+
+logger = logging.getLogger(__name__)
+
+loops = {
+    # Store whatever loop the user has on their thread the client was created in.
+    # We can use this to schedule tasks for their handler/callback code so that any
+    # poor performance in their provided code doesn't slow down the client.
+    "USER_LOOP": asyncio.get_event_loop(),
+    "CLIENT_INTERNAL_LOOP": None,
```

Comment:
You need to defer the call to `get_event_loop`. Calling this here mean that the event loop is created when this file gets imported before any customer code has a chance to run.  

Here's one bug that this introduces: We sneakily create the event loop before the user has a chance to set a custom event loop policy.  

```
from azure.iot.device import AzureIoTDeviceClient    # creates an event loop according to the default policy
import asyncio

asyncio.set_event_loop_policy(my_custom_policy)    # sets the policy after the current thread's loop has already been created

asyncio.run(my_main_coroutine)    # user expects code to run on loop created by my_custom_policy, but it's running on the default loop.
```

This kind of bug is __extremely__ hard to track down because you don't expect the unrelated `import` statement at the top of the file to change the behavior of code in such a subtly destructive way.

Reply:
Yeah, this is why I didn't complete the functionality - this is gonna take a bit more work.

---
In reply to: [480443287](https://github.com/Azure/azure-iot-sdk-python/pull/651#discussion_r480443287) [](ancestors = 480443287)